drivers: sdhc: added atmel hsmci driver

plinnie opened the pull request:
This commit adds support for the ATMEL HSMCI peripheral for the SAM4E MCU series, enabling native SD card support.

pdgendt commented on boards/arm/sam4e_xpro/sam4e_xpro-pinctrl.dtsi:
```diff
@@ -94,4 +94,16 @@
+	hsmci_default: hsmci_default {
+		group1 {
+			pinmux = <PA28C_HSMCI_MCCDA>,
+				<PA29C_HSMCI_MCCK>,
+				<PA30C_HSMCI_MCDA0>,
+				<PA31C_HSMCI_MCDA1>,
+				<PA26C_HSMCI_MCDA2>,
+				<PA27C_HSMCI_MCDA3>;
```
Nit: fix indentation
```suggestion
				 <PA29C_HSMCI_MCCK>,
				 <PA30C_HSMCI_MCDA0>,
				 <PA31C_HSMCI_MCDA1>,
				 <PA26C_HSMCI_MCDA2>,
				 <PA27C_HSMCI_MCDA3>;
```

pdgendt reviewed (changes requested):
Besides the compliance issues and comments, the PR should be split into multiple commits:
- Adding driver
- Updating board
- Updating sample

vvanbeveren committed:
driver: sdhc: added atmel SAM4E hsmci driver

This commit adds support for the ATMEL HSMCI peripheral
for the SAM4E MCU series, enabling native SD card support.

Signed-off-by: Vincent van Beveren <v.van.beveren@nikhef.nl>

vvanbeveren committed:
boards: sam4_xpro: add HSMCI support

Adds support for the HSMCI peripheral enabling SD card access
on the SAM4E eXplained pro board

Signed-off-by: Vincent van Beveren <v.van.beveren@nikhef.nl>

danieldegrasse commented on drivers/sdhc/Kconfig.sam_hsmci:
```diff
@@ -0,0 +1,8 @@
+
+config SAM_HSMCI
+	bool "ATMEL SAM HSMCI driver"
+	select SDHC_SUPPORTS_NATIVE_MODE
```
This Kconfig should be declared like the following:

```
config SAM_HSMCI
     bool "ATMEL SAM HSMCI driver"
     default y
     depends on DT_HAS_AMTEL_SAM_HSMCI_ENABLED
     select SDHC_SUPPORTS_NATIVE_MODE
     select PINCTRL
     help
       Enable the ATMEL SAM HSMCI MMC/SD card driver.
```

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,537 @@
+
+#define _HSMCI_DEFAULT_TIMEOUT 5000
+#define _HSMCI_MAX_FREQ        (SOC_ATMEL_SAM_MCK_FREQ_HZ >> 1)
+#define _HSMCI_MIN_FREQ        (_HSMCI_MAX_FREQ / 0x200) // only valid when MCI has ODD bit
```
General comment: please use C89 `/* */` style comments only

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,537 @@
+	[SD_RSP_TYPE_R2] = HSMCI_CMDR_RSPTYP_136_BIT,
+	[SD_RSP_TYPE_R3] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R4] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R5] = 0 /*?*/,
```
R5 and R5b are used for SDIO- If these are not supported by this peripheral, maybe add a comment here making that clear?

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,537 @@
+
+static int sam_hsmci_reset(const struct device *dev)
+{
+	LOG_DBG("sam_hsmci_reset()");
```
Minor nit here: why not use `LOG_DBG("%s()", __func__)`?

Also, please move this LOG statement after variable declarations

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,537 @@
+	memset(props, 0, sizeof(*props));
+	props->f_max = _HSMCI_MAX_FREQ;
+	props->f_min = _HSMCI_MIN_FREQ;
+	// true not working yet
```
Update this comment to be a bit more clear please- it seems to indicate that high speed support is not yet functional with the peripheral

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,537 @@
+
+static int sam_hsmci_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	LOG_DBG("sam_hsmci_set_io(clock=%d, bus_width=%d, timing=%d, mode=%d)", ios->clock,
```
Same question here- using `__func__` macro for the function name might make this message a bit more future proof. Also, please place this LOG statement after variable declarations.

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,521 @@
+		return ret;
+	}
+	/* Enable module's clock */
+	(void)clock_control_on(SAM_DT_PMC_CONTROLLER, (clock_control_subsys_t)&config->clock_cfg);
```
Unclear where this definition is coming from, but should this not be pulled from the HSMCI's devicetree entry, via a PHANDLE? Take a look at how other drivers interact with their clock controller device pointers: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sdhc/imx_usdhc.c#L256

danieldegrasse replied:
Ok, so this definition comes out as `DEVICE_DT_GET(DT_NODELABEL(pmc))`. Typically the `DT_INST_CLOCKS_CTLR` macro would be used to get a device pointer to the clock controller for this device instance, but if this is the way atmel drivers are managing clocks then I am ok following the convention here.

plinnie replied:
I'm also following convention here.

danieldegrasse commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,521 @@
+
+	/* init carrier detect (if set) */
+	if (config->carrier_detect.port != NULL) {
+		if (!device_is_ready(config->carrier_detect.port)) {
```
You can use `gpio_is_ready_dt` here

danieldegrasse commented on samples/subsys/fs/fat_fs/boards/sam4e_xpro.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_DISK_DRIVER_SDMMC=y
+CONFIG_SDHC=y
+CONFIG_SAM_HSMCI=y
```
You should not need to explicitly enable the driver using this overlay, see my comment about how the `SAM_HSMCI` Kconfig should be defined

danieldegrasse commented on samples/subsys/fs/fat_fs/boards/sam4e_xpro.overlay:
since this overlay will be required for any SDHC test, could you just define the SDMMC device at the board level?

plinnie replied:
Is the 
```
  mmc {
      compatible = "zephyr,sdmmc-disk";
      status = "okay";
  };
```
statement not specific to the sdmmc disk device for the fat-fs example? It is not part of the board it self.

danieldegrasse replied:
For in tree boards, I have defined the SDMMC device at the board level, even when only an SD slot is present on the board. I prefer this approach as it avoids the need for an overlay being added to every test that is going to support a given SDHC on a board, therefore meaning that new tests or samples that rely on the SDMMC device being present should "just work" on the board.

plinnie replied:
Alright, I have modified it like this. It does make the PR a bit simpler.

danieldegrasse commented on tests/drivers/disk/disk_performance/boards/sam4e_xpro.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_DISK_DRIVER_SDMMC=y
+CONFIG_SDHC=y
+CONFIG_SAM_HSMCI=y
```
Same comment, `CONFIG_SAM_HSMCI` is likely not needed.

pdgendt commented on samples/subsys/fs/fat_fs/boards/sam4e_xpro.overlay:
```diff
@@ -0,0 +1,6 @@
+    mmc {
+        compatible = "zephyr,sdmmc-disk";
+        status = "okay";
+    };
```
Nit: Use tabs for indentation.

pdgendt commented on samples/subsys/fs/fat_fs/boards/sam4e_xpro.overlay:
```diff
@@ -0,0 +1,6 @@
+&hsmci {
```
Not sure why there isn't a compliance issue reported, but there should be a SPDX license header.

pdgendt commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,521 @@
+		ios->bus_width, ios->timing, ios->bus_mode);
+
+	if (ios->clock > 0) {
+		int div_val = SOC_ATMEL_SAM_MCK_FREQ_HZ / ios->clock - 2;
```
You should probably get the current clock frequency using
```c
	uint32_t frequency;
	ret = clock_control_get_rate(SAM_DT_PMC_CONTROLLER,
				     (clock_control_subsys_t)&config->clock_cfg,
				     &frequency);
```

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+
+static void sam_hsmci_send_clocks(Hsmci *hsmci)
+{
+	hsmci->HSMCI_MR &= ~(HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF | HSMCI_MR_FBYTE);
+	hsmci->HSMCI_ARGR = 0;
+	hsmci->HSMCI_CMDR =
+		HSMCI_CMDR_RSPTYP_NORESP | HSMCI_CMDR_SPCMD_INIT | HSMCI_CMDR_OPDCMD_OPENDRAIN;
+	while (!(hsmci->HSMCI_SR & HSMCI_SR_CMDRDY)) {
+		;
+	}
+	hsmci->HSMCI_MR |= HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF;
```
Wrong style
```diff
+	hsmci->HSMCI_MR &= ~(HSMCI_MR_WRPROOF
+                           | HSMCI_MR_RDPROOF
+                           | HSMCI_MR_FBYTE);
	hsmci->HSMCI_ARGR = 0;
	hsmci->HSMCI_CMDR = HSMCI_CMDR_RSPTYP_NORESP
+                         | HSMCI_CMDR_SPCMD_INIT
+                         | HSMCI_CMDR_OPDCMD_OPENDRAIN;
	while (!(hsmci->HSMCI_SR & HSMCI_SR_CMDRDY)) {
		;
	}
	hsmci->HSMCI_MR |= HSMCI_MR_WRPROOF
+                        | HSMCI_MR_RDPROOF;
```

Please, review this on the whole code.

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
```
```diff
		if (sr & (HSMCI_SR_UNRE
		         | HSMCI_SR_OVRE
		         | HSMCI_SR_DTOE
		         | HSMCI_SR_DCRCE)) {
```

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+
+		hsmci->HSMCI_MR |= HSMCI_MR_PDCMODE;
+
+		hsmci->HSMCI_BLKR =
+			HSMCI_BLKR_BLKLEN(sd_data->block_size) | HSMCI_BLKR_BCNT(sd_data->blocks);
```
```diff
		hsmci->HSMCI_BLKR = HSMCI_BLKR_BLKLEN(sd_data->block_size)
				  | HSMCI_BLKR_BCNT(sd_data->blocks);
```

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+
+	hsmci->HSMCI_PTCR = HSMCI_PTCR_RXTDIS | HSMCI_PTCR_TXTDIS;
+
+	struct sdhc_command cmd = {
+		.opcode = SD_STOP_TRANSMISSION, .arg = 0, .response_type = SD_RSP_TYPE_NONE};
```
```
	struct sdhc_command cmd = {.opcode = SD_STOP_TRANSMISSION,
				   .arg = 0,
				   .response_type = SD_RSP_TYPE_NONE};
```

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+		return -EBUSY;
+	}
+
+#ifdef _HSMCI_PWRSAVE
```
dito

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+		}
+	} while (ret != 0 && (cmd->retries-- > 0));
+
+#ifdef _HSMCI_PWRSAVE
```
dito

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+				if (!sam_hsmci_card_busy(dev)) {
+					break;
+				}
+				/* Wait 125us before polling again */
```
the busy wait is clear, you can drop this comment

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+			hsmci->HSMCI_MR |= HSMCI_MR_FBYTE;
+		}
+
+		hsmci->HSMCI_MR |= HSMCI_MR_PDCMODE;
```
There is no PDC driver at moment. Can you clarify?
I believe this is the only bit that not allow use of this driver with all SoC's.

plinnie replied:
The PDC (Peripheral DMA Controller) is a DMA controller which is combined with the specific peripheral in the same hardware 'IP'. Due to this the SAM4E has PDC registers appended in the memory region of the HSMCI. So, the PDC usage is inside of the driver, like here:

https://github.com/zephyrproject-rtos/zephyr/blob/800cad694002b76ea4016feb04b1f6b678a7887e/drivers/sdhc/sam_hsmci.c#L423-L430

All registers here, except for PTCR, are PDC registers. 

Other parts of AMTEL (I have to look  up which) also have the HSMCI peripheral, but use a generic DMA peripheral to execute the actual DMA transfer. In that case you would need to actually support the DMA controller specific to that part. Which is something I can't test, so will be difficult for me to write.

However, you do have a point, as it should also perfectly possible to not use the PDC, if not available. This will be a bit slower, but should allow support for a more broad range of parts, and I can test this on my own chip. I will see if I can incorporate this into the driver.

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+#include <zephyr/drivers/clock_control/atmel_sam_pmc.h>
+
+#ifndef HSMCI_MR_PDCMODE
+#error "Sorry, only SAM4E/SAM4S are supported at the moment"
```
Are you sure that we can not drop this and make it compatible with remaining devices?
`hsmci->HSMCI_MR |= HSMCI_MR_PDCMODE;`

if not
```diff
+#error "The driver currently supports SAM4 SoC's"
```

plinnie replied:
Please see my previous comment.

nandojve reviewed (changes requested):
Hi @plinnie,
Thank you for this contribution. I have a few styles to mention.

The relevant part is understand if we realy need HSMCI_MR_PDCMODE and props can be defined at devicetree.

plinnie commented:
I'm a bit busy, but still working on the PR. I have removed the requirement of PDC mode, so it should work in more general cases. I hope to have the next PR iteration ready next week

plinnie commented:
Okay, everything seems to be in order now. Please review.

plinnie commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+
+LOG_MODULE_REGISTER(hsmci, CONFIG_SDHC_LOG_LEVEL);
+
+#define _HSMCI_DEFAULT_TIMEOUT 5000
+#define _HSMCI_MAX_FREQ        (SOC_ATMEL_SAM_MCK_FREQ_HZ >> 1)
+#define _HSMCI_MIN_FREQ        (_HSMCI_MAX_FREQ / 0x200)
+#define _HSMCI_PWRSAVE         7
```
I run clang format before each save. It doesn't like it.

plinnie commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+#define _HSMCI_MIN_FREQ        (_HSMCI_MAX_FREQ / 0x200)
+#define _HSMCI_PWRSAVE         7
+
+#define _HSMCI_SR_ERR                                                                              \
+	(HSMCI_SR_RINDE | HSMCI_SR_RDIRE | HSMCI_SR_RCRCE | HSMCI_SR_RENDE | HSMCI_SR_RTOE |       \
+	 HSMCI_SR_DCRCE | HSMCI_SR_DTOE | HSMCI_SR_CSTOE | HSMCI_SR_OVRE | HSMCI_SR_UNRE)
```
and this, it reverts

plinnie commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+	 HSMCI_SR_DCRCE | HSMCI_SR_DTOE | HSMCI_SR_CSTOE | HSMCI_SR_OVRE | HSMCI_SR_UNRE)
+
+static const uint8_t _resp2size[] = {
+	[SD_RSP_TYPE_NONE] = HSMCI_CMDR_RSPTYP_NORESP,
+	[SD_RSP_TYPE_R1] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R1b] = HSMCI_CMDR_RSPTYP_R1B,
+	[SD_RSP_TYPE_R2] = HSMCI_CMDR_RSPTYP_136_BIT,
+	[SD_RSP_TYPE_R3] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R4] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R5] = 0 /* SDIO not supported */,
+	[SD_RSP_TYPE_R5b] = 0 /* SDIO not supported */,
+	[SD_RSP_TYPE_R6] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R7] = HSMCI_CMDR_RSPTYP_48_BIT,
```
and this also

plinnie commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+	props->f_max = _HSMCI_MAX_FREQ;
+	props->f_min = _HSMCI_MIN_FREQ;
+	/* high-speed not working yet due to limitations of the SDHC sm */
+	props->host_caps.high_spd_support = false;
+	props->power_delay = 500;
+	props->is_spi = false;
+	props->max_current_330 = 4;
```
No, they are device specific

plinnie commented:
Please hold. I have to yet verify it passes the hardware test. Will do tomorrow.

plinnie commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,538 @@
+	hsmci->HSMCI_CR = HSMCI_CR_SWRST;
+	hsmci->HSMCI_CR = HSMCI_CR_PWSDIS;
+	hsmci->HSMCI_CR = HSMCI_CR_MCIEN;
+#ifdef _HSMCI_PWRSAVE
```
Converted this to Kconfig options

nandojve commented:
Hi @plinnie ,

I've a SAM4E_XPRO and a SAM_V71_XULT. Once you validate I'll check from my side.

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,673 @@
+			LOG_DBG("PDC sr 0x%08x error", sr);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_TXBUFE));
```
Changing to HSMCI_SR_XFRDONE I can build for SAM4E and SAMV71. SAM4E still passing on tests and SAMV71 fail at tests/drivers/disk/disk_access first becase of 

```diff
-static uint8_t scratch_buf[2][SECTOR_COUNT4 * SECTOR_SIZE + 1];
+static uint8_t scratch_buf[2][(SECTOR_COUNT4 + 1) * SECTOR_SIZE];
```

After that fix I got this:

```shell
*** Booting Zephyr OS build zephyr-v3.4.0-2535-g62f4249080c8 ***
Running TESTSUITE disk_driver
===================================================================
Disk reports 3911680 sectors
Disk reports sector size 512
START - test_read
Testing reads of 8 sectors
Testing reads of 1 sectors
Testing reads of 29 sectors
Testing reads of 31 sectors

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/disk/disk_access/src/main.c:196: disk_driver_test_read: (scratch_buf[1] not equal )
Multiple reads mismatch
 FAIL - test_read in 0.072 seconds
===================================================================
START - test_write
Testing writes of 8 sectors

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/disk/disk_access/src/main.c:142: write_sector_checked: (wbuf not equal to rbuf)
Read data did not match data written to disk
 FAIL - test_write in 0.027 seconds
===================================================================
TESTSUITE disk_driver failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -   0.00% [disk_driver]: pass = 0, fail = 2, skip = 0, total = 2 duration = 0.099 seconds
 - FAIL - [disk_driver.test_read] duration = 0.072 seconds
 - FAIL - [disk_driver.test_write] duration = 0.027 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```

Any idea about possible problem?
BTW, this is not a block!

nandojve commented on drivers/sdhc/sam_hsmci.c:
```diff
@@ -0,0 +1,673 @@
+							     HSMCI_SR_DTOE | HSMCI_SR_DCRCE));
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_RXBUFF));
```
I dropped all this **} while (!(sr & HSMCI_SR_RXBUFF));** block to build for SAMV71 and SAM4E still passing on tests.
Can you check if this is really necessary?

plinnie replied:
You are right. These registers are for PDC, which the V71 does not have. The SAM4E should use it when using PDC. I have wrapped these functions in ifdefs. I'll hope to update the PR today.

plinnie replied:
@nandojve It works from my side. Should work on the V71 also. Can you check please?

nandojve reviewed (commented):
Hi @plinnie ,

I tested both SAM4E and SAMV71. See my comments and a possible bug on a test.
The SAMV71 is not mandatory to us move this PR forward but it will be very nice if we can found a ground for both MCUs.

I think you already have all info about driver and you can start to apply the code style changes.
It may take a few interactions to fix those cosmetics but after that it is ready to be merged.

plinnie commented:
> I think you already have all info about driver and you can start to apply the code style changes. It may take a few interactions to fix those cosmetics but after that it is ready to be merged.


Okay, I will update the style, but maybe we should also create an ATMEL specific clang format, since it undoes all my work each time I reformat the document🙄.

nandojve reviewed (changes requested):
Hi @plinnie ,

I tested on SAMV71 and SAM4E. In the SAMV71 the disk_access still not work but it is not a block so the overall LGTM!

I found a few small cosmetics. BTW, I'm not sure if you already updated the style : )

> Okay, I will update the style, but maybe we should also create an ATMEL specific clang format, since it undoes all my work each time I reformat the code 🙄.

I'm working on a clang-format to make it available at hal_atmel. Thank you for the suggestion!

plinnie commented:
> Hi @plinnie ,
> 
> I tested on SAMV71 and SAM4E. In the SAMV71 the disk_access still not work but it is not a block so the overall LGTM!
> 
> I found a few small cosmetics. BTW, I'm not sure if you already updated the style : )
> 
> > Okay, I will update the style, but maybe we should also create an ATMEL specific clang format, since it undoes all my work each time I reformat the code 🙄.
> 
> I'm working on a clang-format to make it available at hal_atmel. Thank you for the suggestion!

Thanks @nandojve  .. so, I've processed your comments. The PR should be OK now

danieldegrasse reviewed (commented):
One (non blocking) comment. Driver otherwise looks good to me
